Add initial inotify exporter implementation

stephen-soltesz opened the pull request:
This change adds initial .travis.yml with goveralls configuration for the watch package, and a minimal inotify_exporter command.

The command inotify_exporter is suitable for bundling with an existing experiment such as NDT to monitor the file creation, deletion, and closewrite rate for various file types.

<!-- Reviewable:start -->
---
This change is [<img src="https://reviewable.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/m-lab/inotify-exporter/2)
<!-- Reviewable:end -->

stephen-soltesz commented:
Review status: 0 of 4 files reviewed at latest revision, 1 unresolved discussion.

---

*[.travis.yml, line 12 at r1](https://reviewable.io:443/reviews/m-lab/inotify-exporter/2#-KxnJBVn-8R9PNASGR0T:-KxnJBVoprQAdEYdetzs:b-7ofhye) ([raw file](https://github.com/m-lab/inotify-exporter/blob/3e6f5e9833fd059cdc172b090c85b76e677399a0/.travis.yml#L12)):*
> ```YAML
> 
> before_install:
> - go get github.com/mattn/goveralls
> ```

+enhancement 

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/m-lab/inotify-exporter/2)*
<!-- Sent from Reviewable.io -->

stephen-soltesz commented:
Review status: 0 of 4 files reviewed at latest revision, 2 unresolved discussions.

---

*[cmd/inotify_exporter/main.go, line 122 at r2](https://reviewable.io:443/reviews/m-lab/inotify-exporter/2#-KxnOvV64NUcvzljTR1W:-KxnOvV71H4l2ODewmtm:b-9v3c7m) ([raw file](https://github.com/m-lab/inotify-exporter/blob/cea8ceab81d8173812f251a46f934c216e6b8cd0/cmd/inotify_exporter/main.go#L122)):*
> ```Go
> 	http.Handle("/metrics", promhttp.Handler())
> 
> 	// TODO: use a port that won't conflict with registered exporter ports.
> ```

I don't think we'll ever register this port https://github.com/prometheus/prometheus/wiki/Default-port-allocations since this exporter is not generic enough to be of general interest to others. But, we should chose one that we don't expect to conflict.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/m-lab/inotify-exporter/2)*
<!-- Sent from Reviewable.io -->

gfr10598 commented:
<img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/>

---

Reviewed 3 of 4 files at r1, 1 of 1 files at r2.
Review status: all files reviewed at latest revision, 9 unresolved discussions.

---

*[cmd/inotify_exporter/main.go, line 77 at r2](https://reviewable.io:443/reviews/m-lab/inotify-exporter/2#-KxnaPf_zfcEpbqU5dHA:-KxnaPf_zfcEpbqU5dHB:b-62okym) ([raw file](https://github.com/m-lab/inotify-exporter/blob/cea8ceab81d8173812f251a46f934c216e6b8cd0/cmd/inotify_exporter/main.go#L77)):*
> ```Go
> }
> 
> // onEvent processes a single inotify event. Only file events are considered.
> ```

This comment seems to describe how it is used, rather than what it does.  Does the prefix matter?  Does it ignore non file events?  Or should it only be called with file events?

---

*[watch/watch.go, line 22 at r1](https://reviewable.io:443/reviews/m-lab/inotify-exporter/2#-KxnJ2MgX_neZ3F2cpK6:-KxnJ2MgX_neZ3F2cpK7:b32v57) ([raw file](https://github.com/m-lab/inotify-exporter/blob/3e6f5e9833fd059cdc172b090c85b76e677399a0/watch/watch.go#L22)):*
> ```Go
> 	monthPattern = `0[1-9]|1[0-2]`
> 	dayPattern   = `0[1-9]|[12][0-9]|3[0-1]`
> 	datePattern  = regexp.MustCompile(yearPattern + "/" + monthPattern + "/" + dayPattern)
> ```

+1

---

*[watch/watch.go, line 28 at r1](https://reviewable.io:443/reviews/m-lab/inotify-exporter/2#-KxnXGztfam_Bu-fpYWA:-KxnXGztfam_Bu-fpYWB:b-ypidwn) ([raw file](https://github.com/m-lab/inotify-exporter/blob/3e6f5e9833fd059cdc172b090c85b76e677399a0/watch/watch.go#L28)):*
> ```Go
> // InCloseWrite events on baseDir and subdirectories. The watch will remain
> // active until the caller sends a value to the stop channel.
> func DirRecursively(baseDir string, stop <-chan bool,
> ```

use:  stop <-chan struct{}
(https://dave.cheney.net/2013/04/30/curious-channels)

---

*[watch/watch.go, line 72 at r1](https://reviewable.io:443/reviews/m-lab/inotify-exporter/2#-Kxn_w0GhXfXyZ3y-PA6:-Kxn_w0GhXfXyZ3y-PA7:b-qjvi7s) ([raw file](https://github.com/m-lab/inotify-exporter/blob/3e6f5e9833fd059cdc172b090c85b76e677399a0/watch/watch.go#L72)):*
> ```Go
> 		t := time.Now().UTC()
> 		log.Printf("%s %s %s\n", t, ev.Event(), ev.Path())
> 		// Process all other events, since this is a recursive watch.
> ```

This comment is hard to understand.

---

*[watch/watch.go, line 74 at r1](https://reviewable.io:443/reviews/m-lab/inotify-exporter/2#-KxnPqKUxN3dMIDcF6K4:-KxnPqKUxN3dMIDcF6K5:b-hsct7c) ([raw file](https://github.com/m-lab/inotify-exporter/blob/3e6f5e9833fd059cdc172b090c85b76e677399a0/watch/watch.go#L74)):*
> ```Go
> 		// Process all other events, since this is a recursive watch.
> 		onEvent(t, ev, shortPath)
> ```

remove blank line?

---

*[watch/watch.go, line 14 at r2](https://reviewable.io:443/reviews/m-lab/inotify-exporter/2#-KxncypaNcmg6xK-eyHv:-KxncypaNcmg6xK-eyHw:b-89u9oo) ([raw file](https://github.com/m-lab/inotify-exporter/blob/cea8ceab81d8173812f251a46f934c216e6b8cd0/watch/watch.go#L14)):*
> ```Go
> 	"golang.org/x/sys/unix"
> 
> 	"github.com/rjeczalik/notify"
> ```

Add TODO to vendor this?

---

*[watch/watch.go, line 67 at r2](https://reviewable.io:443/reviews/m-lab/inotify-exporter/2#-KxndAGgmXsUxv5GcNF-:-KxndAGgmXsUxv5GcNF0:b-eft5q8) ([raw file](https://github.com/m-lab/inotify-exporter/blob/cea8ceab81d8173812f251a46f934c216e6b8cd0/watch/watch.go#L67)):*
> ```Go
> 		// Only accept valid YYYY/MM/DD paths.
> 		if !isValidPath(shortPath) {
> 			continue
> ```

Maybe invert this, and put the remainder inside?

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/m-lab/inotify-exporter/2#-:-Kxndcx1r3YA_hvkaRW9:bnfp4nl)*
<!-- Sent from Reviewable.io -->

stephen-soltesz commented:
Thank you!

---

Review status: 1 of 4 files reviewed at latest revision, 10 unresolved discussions.

---

*[cmd/inotify_exporter/main.go, line 77 at r2](https://reviewable.io:443/reviews/m-lab/inotify-exporter/2#-KxnaPf_zfcEpbqU5dHA:-KxoQw9tHaP1KRLhS69n:bx1kppd) ([raw file](https://github.com/m-lab/inotify-exporter/blob/cea8ceab81d8173812f251a46f934c216e6b8cd0/cmd/inotify_exporter/main.go#L77)):*
<details><summary><i>Previously, gfr10598 (Gregory Russell) wrote…</i></summary><blockquote>

This comment seems to describe how it is used, rather than what it does.  Does the prefix matter?  Does it ignore non file events?  Or should it only be called with file events?
</blockquote></details>

Good point. At one point this wasn't a user-provided function. I've updated the function name to be more meaningful and updated the docstring to match it's behavior.

---

*[watch/watch.go, line 28 at r1](https://reviewable.io:443/reviews/m-lab/inotify-exporter/2#-KxnXGztfam_Bu-fpYWA:-KxoMz_lzYOSZSXUFXzh:b5qj149) ([raw file](https://github.com/m-lab/inotify-exporter/blob/3e6f5e9833fd059cdc172b090c85b76e677399a0/watch/watch.go#L28)):*
<details><summary><i>Previously, gfr10598 (Gregory Russell) wrote…</i></summary><blockquote>

use:  stop <-chan struct{}
(https://dave.cheney.net/2013/04/30/curious-channels)

</blockquote></details>

Aah. That is much better. Done.

---

*[watch/watch.go, line 72 at r1](https://reviewable.io:443/reviews/m-lab/inotify-exporter/2#-Kxn_w0GhXfXyZ3y-PA6:-KxoOKar-HiraZStmIj5:bkd106l) ([raw file](https://github.com/m-lab/inotify-exporter/blob/3e6f5e9833fd059cdc172b090c85b76e677399a0/watch/watch.go#L72)):*
<details><summary><i>Previously, gfr10598 (Gregory Russell) wrote…</i></summary><blockquote>

This comment is hard to understand.
</blockquote></details>

Updated the comment to be a little clearer.

---

*[watch/watch.go, line 74 at r1](https://reviewable.io:443/reviews/m-lab/inotify-exporter/2#-KxnPqKUxN3dMIDcF6K4:-KxoOO_8x1LvXuw0xCEL:b-896fix) ([raw file](https://github.com/m-lab/inotify-exporter/blob/3e6f5e9833fd059cdc172b090c85b76e677399a0/watch/watch.go#L74)):*
<details><summary><i>Previously, gfr10598 (Gregory Russell) wrote…</i></summary><blockquote>

remove blank line?
</blockquote></details>

Done.

---

*[watch/watch.go, line 14 at r2](https://reviewable.io:443/reviews/m-lab/inotify-exporter/2#-KxncypaNcmg6xK-eyHv:-KxoJQ6Hvy-0T6eCoQsd:b-896fix) ([raw file](https://github.com/m-lab/inotify-exporter/blob/cea8ceab81d8173812f251a46f934c216e6b8cd0/watch/watch.go#L14)):*
<details><summary><i>Previously, gfr10598 (Gregory Russell) wrote…</i></summary><blockquote>

Add TODO to vendor this?
</blockquote></details>

Done.

---

*[watch/watch.go, line 67 at r2](https://reviewable.io:443/reviews/m-lab/inotify-exporter/2#-KxndAGgmXsUxv5GcNF-:-KxoRCDXMfDT7-CW_97s:b-896fix) ([raw file](https://github.com/m-lab/inotify-exporter/blob/cea8ceab81d8173812f251a46f934c216e6b8cd0/watch/watch.go#L67)):*
<details><summary><i>Previously, gfr10598 (Gregory Russell) wrote…</i></summary><blockquote>

Maybe invert this, and put the remainder inside?
</blockquote></details>

Done.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/m-lab/inotify-exporter/2#-:-KxoRTTFPBh5CXyMf1Da:b-9755fn)*
<!-- Sent from Reviewable.io -->
